manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java

```diff
@@ -0,0 +1,639 @@
+    //            A state change happened where the state we are looking for was turned on.
+    //            We want the currentDp.
+
+
+                if(currentDpTest && !previousDpTest){
+                    StateChangeAssetDatapoint = currentDp;
+                    break;
+                }
+            }
+
+            if (StateChangeAssetDatapoint != null){
+                if (!pred.value.asPredicate(timerService::getCurrentTimeMillis).test(StateChangeAssetDatapoint.getValue())){
+                    throw new Exception("Found state change datapoint failed predicate");
+                }
+            }else{
+                throw new Exception("Couldn't find asset state change value");
+            }
+
+        }catch (Exception e){
+            getLogger().warning(e.getMessage());
+            return Optional.empty();
+        }
+
+        //BUG: The Timestamp of an asset datapoint is NOT attribute.getTimestamp()!
+        //This makes it much harder to relate datapoints to relate a duration to the time-distance between two datapoints.
+        //To make this easier, we are going to use some time-padding of 1 second from the gap. It's going to be sufficient to cover some edge-cases.
+
+        //The AssetStateDuration will have a startTime of the found state-change variable, and the endDate will be the previousValue's timestamp, that being the end of the state duration
+        Attribute<?> tripAttr = new Attribute<>("LastTripStartedAndEndedAt", ValueType.ASSET_STATE_DURATION, new AssetStateDuration(
+                new Timestamp(StateChangeAssetDatapoint.getTimestamp()),
+                new Timestamp(previousValue.getTimestamp().get())
+        ));
+
+        tripAttr.setTimestamp(StateChangeAssetDatapoint.getTimestamp());
+
+        return Optional.of(tripAttr);
+    }
+
+    private String getParameterFileString() {
+        try {
+            return Files.readString(DeviceParameterPath);
+        } catch (IOException e) {
+            getLogger().warning("Couldn't find FMC003.json, couldn't parse parameters");
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Updates the {@link Asset} passed, with the {@link AttributeMap} passed.
+     * I have a very big feeling that the way this is done is wrong.
+     *
+     * @param asset The asset to be updated.
+     * @param attributes The attributes to be upserted to the Attribute.
+     * @param topic The topic to which the MQTT payload was sent.
+     * @param connection The connection on which the payload was sent.
+     */
+    private void UpdateAsset(Asset<?> asset, AttributeMap attributes, Topic topic, RemotingConnection connection) {
+
+        String imei = asset.getAttribute("IMEI").toString();
+
+        getLogger().info("Updating CarAsset with IMEI "+imei);
+        //DONE: If the parameters do not exist, then the update fails,
```

Comment:
Remove these notes when merging I guess? Don't know how important they are?

It's also a kind of general practice to not use too many multiline comments in methods.
Placing a large comment above the function describing it is indeed a good habit,
but in the method itself it should only shortly describe its meaning and possible behavior.

Reply:
Removed a lot of them, but I am keeping the TODOs until I have a very stable (or merge-able) version. When I do, this will be converted to proper entries in a Kanban board